learningequality/studio
# Segment clipboard by channels

According to @khangmach 's awesome mocks!

### Screenshots

![image](https://user-images.githubusercontent.com/159687/41383861-a415b574-6f27-11e8-8339-beef40128428.png)

![image](https://user-images.githubusercontent.com/159687/41383865-abd29f5c-6f27-11e8-8fb3-f967ee3aae95.png)

### Implementation notes

These UI segments are implemented pretty much entirely on the frontend, re-using the views `ClipboardItem` with a nested `ClipboardList` to render the channel nodes and their contents respectively. These views use the models `ContentNodeModel` and `ContentNodeCollection` respectively.

The only modification made to the API / backend is augmenting node data (in `ContentNodeSerializer`) with a `source_channel` dict containing fields `id`, `name`, and `thumbnail_url`. This unfortunately incurs a slight, but not significant, perf hit as it involves a DB query to get a node's source node in the serializer (see comments in code for profiling details). Open to suggestions for optimizing this — I have not put much thought into this at all.

Back to the frontend code, in `views.ClipboardList`, we take the flat list of clipboard content nodes we get from the API, and transform it by grouping together nodes of the same source channel (with the `group_by_channel()` method). We end up with a list of `ContentNodeModel`s, each of which represents a channel node, and has children contained in a `ContentNodeCollection`. Frontend code that add node(s) to the clipboard and copying within it has been updated so that the node(s) are inserted into the appropriate segment, while ensuring API calls aren't affected.

These clipboard segments can be adapted to show arbitrary lists (rather than channels), by changing the `group_by_channel()` method to group by the list the node belongs to, and changing the `ClipboardList#is_segment()` and `ClipboardItem#is_segment()` methods accordingly.

As far as I can tell, there's no significant frontend perf hit (at least in the screenshots shown above).

### Test plan

- Copy some content items from different channels to the clipboard by using the "Make a copy" button, icon, or dragging it in
- See that it all works w/o errors in the JS console, and it gets segmented into channels like the screenshots above.
- Click "make a copy" on one of the clipboard items and see that it creates a copy of the item shown in the bottom of the channel segment
- Click "Delete" on some nodes and see that the item is removed
- Drag an item out of the clipboard onto the main tree, and see that it gets added into the main tree and removed from the clipboard
- Ensure there are no other JS errors in the JS console

Known issue: Whenever a new node gets added to the clipboard, all channels get expanded again (even if some were collapsed). There is a TODO comment in the code to fix it, and I'm happy to address that in this PR if desirable!

### Original commit names (now squashed)

- Get segmenting to work using ClipboardList
- Support adding nodes to clipboard
- Add `source_channel` attribute to the JSON we send down
- Fix resetting scroll position; get drag out to work; item deletion
- Style it according to spec! Mostly except w/o colors
- Add accent bars to clipboard styling
- Clean up code -- get it ready for review; fix drag/drop issues
- Fix some whitespace and remove a comment
